### DISCUSSION TITLE
v2.13.0

### DISCUSSION THREAD

**Discussion by SevereCloud** [SevereCloud/vksdk#181] [General]
### Поддержка MessagePack и zstd

> Результат перехода с gzip (JSON) на zstd (msgpack):
>
> - в 7 раз быстрее сжатие (–1 мкс);
> - на 10% меньше размер данных (8 Кбайт вместо 9 Кбайт);
> - продуктовый эффект не статзначимый :(
>
> [Как мы отказались от JPEG, JSON, TCP и ускорили ВКонтакте в два раза](https://habr.com/ru/company/vk/blog/594633/)

VK API способно возвращать ответ в виде [MessagePack](https://msgpack.org/). Это эффективный формат двоичной сериализации, похожий на JSON, только быстрее и меньше по размеру.

ВНИМАНИЕ, C MessagePack НЕКОТОРЫЕ МЕТОДЫ МОГУТ ВОЗВРАЩАТЬ СЛОМАННУЮ КОДИРОВКУ.

Для сжатия, вместо классического gzip, можно использовать [zstd](https://github.com/facebook/zstd). Сейчас vksdk поддерживает zstd без словаря. Если кто знает как получать словарь, [отпишитесь сюда](https://github.com/SevereCloud/vksdk/issues/180).

```go
vk := api.NewVK(os.Getenv("USER_TOKEN"))

method := "store.getStickersKeywords"
params := api.Params{
	"aliases":       true,
	"all_products":  true,
	"need_stickers": true,
}

r, err := vk.Request(method, params) // Content-Length: 44758 byte
if err != nil {
	log.Fatal(err)
}
log.Println("json:", len(r)) // json: 814231 byte

vk.EnableMessagePack() // enable MessagePack
vk.EnableZstd() // enable zstd

r, err = vk.Request(method, params) // Content-Length: 35755 byte
if err != nil {
	log.Fatal(err)
}
log.Println("msgpack:", len(r)) // msgpack: 650775 byte
```

0ed3f0c feat: api support MessagePack
15a6084 feat: api support zstd

### Авторизация используя VK ID

[VK ID](https://static.vk.com/id/) — это единая платформа для авторизации и регистрации пользователей в разных сервисах экосистемы VK.

Для получения данных о пользователя, с фронтенда [передайте](https://platform.vk.com/?p=DocsDashboard&docs=install-sdk_install-sdk-web) `payload`

```ts
// В onAuthData callback-функции в аргументе data будет содержаться
// информация об авторизации
const onAuthData = (data:VKUserVisibleAuthResult) => {
  if (data.provider === "vk") {
    fetch('/accessToken', {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(data.payload)
    });
  }

  console.log(data);
}
```

На бэкенде используйте json декодер.

```go
func accessToken(w http.ResponseWriter, req *http.Request) {
	var payload vkid.SilentAuthPayload
	err := json.NewDecoder(req.Body).Decode(&payload)
	...
}
```

Чтобы получить данные пользователя с помощью [Silent token](https://platform.vk.com/?p=DocsDashboard&docs=tokens_silent-token) вызовите метод `AuthGetProfileInfoBySilentToken`

```go
p := params.NewAuthGetProfileInfoBySilentTokenBuilder()
p.Add(payload.Token, payload.UUID, "click")

r, err := vk.AuthGetProfileInfoBySilentToken(p.Params)
```

Для получения [Access token](https://platform.vk.com/?p=DocsDashboard&docs=tokens_access-token) вызовите метод `vk.AuthExchangeSilentAuthToken`

```go
p := params.NewAuthExchangeSilentAuthTokenBuilder()
p.Token(payload.Token)
p.UUID(payload.UUID)

r, err := vk.AuthExchangeSilentAuthToken(p.Params)
```

aaa7499 feat: add SilentAuthPayload
db2c8e7 feat: add AuthGetProfileInfoBySilentToken method
b286e3d feat: add AuthExchangeSilentAuthToken method

### Разное

c58c2c9 feat: MessagesTemplateElementCarousel add PhotoID (#171)
04b056b Fix parsing {failed: 1} response (by @Delorus)

<hr /><em>This discussion was created from the release <a href='https://github.com/SevereCloud/vksdk/releases/tag/v2.13.0'>v2.13.0</a>.</em>

**Comment by ghost** [comment]
@SevereCloud «ВНИМАНИЕ, C MessagePack НЕКОТОРЫЕ МЕТОДЫ МОГУТ ВОЗВРАЩАТЬ СЛОМАННУЮ КОДИРОВКУ.»

Актуально?

**Reply by SevereCloud** [reply]
Думаю нет, но это не точно
